drivers,server: Cap data read for a seg

jailuthra opened the pull request:
**What does this pull request do? Explain your changes. (required)**
<!-- A clear and concise description of what this pull request does. -->
Cap the data read for a segment to prevent bad segments eating up a lot of memory.

**Specific updates (required)**
<!--- List out all significant updates your code introduces -->
*see the commit history*

Currently the max segment size is set to ~300MB (at 8000kbps using the current max duration of 5minutes)

**How did you test each of these updates (required)**
<!-- A detailed description of how you tested your code changes. Include details of your testing environment, and the tests you ran to see how your change affects other areas of the code, etc. -->

- Manually tested on a local B/OT setup with a smaller limit set (2MB) and a bigger segment (2.4MB). The segment data read was being capped at the defined size.


With Limit Set:
B: `I0404 22:11:10.044429 2542338 broadcast.go:372] Processing segment nonce=17696729379380681746 manifestID=test seqNo=1 dur=2 bytes=2097152`
OT: `[h264 @ 0x7f810403d640] error while decoding MB 60 59, bytestream -12`
Transcoded Rendition Duration: 1.23 s

Without Limit Set:
B: `I0404 22:12:42.698367 2543964 broadcast.go:372] Processing segment nonce=13032173926343571666 manifestID=test seqNo=1 dur=2 bytes=2405084`
OT: *No error*
Transcoded Rendition Duration: 2.7 s

2. Added unit tests for Push API and B<->O serve segment in https://github.com/livepeer/go-livepeer/pull/1827/commits/9d23156dc5aa076b4e8ec0becb5d777eede42eba

**Does this pull request close any open issues?**
<!-- Fixes # -->
Fixes #1523 

**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Read the [contribution guide](./doc/contributing.md)
- [x] `make` runs successfully
- [x] All tests in `./test.sh` pass
- [x] README and other documentation updated
- [x] [Pending changelog](./CHANGELOG_PENDING.md) updated

jailuthra committed:
drivers,server: Cap data read for a seg

jailuthra committed:
common/util: ReadAtMost n bytes from an io.Reader

jailuthra commented on common/util.go:
```diff
@@ -387,3 +392,8 @@ func JoinURL(url, path string) string {
 	return url + path
 }
+
+func ReadAtMost(r io.Reader, n int64) ([]byte, error) {
```
Not logging any warnings or returning an error here for the case when the segment size > `MaxSegSize` (and we end up reading it partially)

We could add some errors/warning around this if we write this function from ground up using manual `Read` calls and increasing buffer gradually if needed - like done in [`ReadAll`'s implementation](https://golang.org/src/io/io.go?s=19249:19287#L616)

darkdarkdragon replied:
I think we should return error from here if not all data was read - because if broadcaster didn't downloaded all segment it doesn't make sense to send it to transcoding - anyway it will be broken segment.

jailuthra replied:
Makes sense on it being a broken segment and needing error log of some sort, will update!

NB: I was able to transcode partial segment on CPU (haven't tested on GPU), so was uncertain if we should still make an attempt at transcoding or not - but for now makes sense to just error out and reconsider later if we see huge segments.

darkdarkdragon replied:
I don't think users expect they segments to be _partially_ transcoded 😄

darkdarkdragon commented on server/mediaserver.go:
```diff
@@ -693,7 +693,7 @@ func (s *LivepeerServer) HandlePush(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 	// we read this unconditionally, mostly for ffmpeg
```
I think this comment is obsolete.

yondonfu commented on common/util.go:
```diff
@@ -26,6 +28,9 @@ import (
 var HTTPTimeout = 8 * time.Second
 
+// Max Segment Size (cap reading HTTP response body at this size)
+const MaxSegSize = 20 * 1024 * 1024 // 20 MiB - 20s of 1080p30fps video @ 8000kbps
```
The max duration of a segment is currently [5 minutes](https://github.com/livepeer/go-livepeer/blob/ffabc358fe3115c81e0fb6fc83f1b6b281194498/server/broadcast.go#L32). Support for segments up to this max duration was added awhile back. I'm not sure how common longer segments as a use case are today (they came up while experimenting with preliminary VOD support + mp4 transcoding), but I think we should set this byte size limit based on the current duration limit.

yondonfu reviewed (changes requested):
After the update to start returning an error if the byte limit is exceeded while reading a request body, let's make sure to test for that error. I recommend small unit test cases as those should make it easy to construct request bodies that exceed the byte limit.

iameli commented:
@yondonfu @jailuthra @darkdarkdragon I feel medium-strongly this should be a configurable parameter with a default value, perhaps `-httpIngestMaxSize=20MiB` or some such. Could use https://github.com/c2h5oh/datasize to make it easy to parse various sizes?

Just feels like the kind of thing I'd like the ability to do in a production setting to unbreak an integration that's ever-so-slightly over the limit.

darkdarkdragon commented:
@iameli we also using this size on the O side. If we make this configurable then we should put this value in discovery protocol so O will clearly advertise its limits - or else B can get unexpected errors.

jailuthra commented:
@iameli If we go with the bigger hardcoded limit of `5minutes` as proposed by @yondonfu above, would we still need to increase it anytime soon-ish in a prod environment?

This can avoid changes on discovery side (for now), and I am definitely in for iterating over this in future!

iameli commented:
Sure, that's reasonable.

jailuthra committed:
common: use MaxDuration for MaxSegSize

yondonfu commented on common/util.go:
```diff
@@ -387,3 +395,15 @@ func JoinURL(url, path string) string {
+	if err == nil && len(b) > n {
+		return nil, errors.New("input bigger than max buffer size")
+	} else {
+		return b, err
```
More idiomatic in Go to do:

```golang
if err == nil && len(b) > n {
    return nil, errors.New("input bigger than max buffer size")
}

return b, err
```

yondonfu commented on common/util.go:
```diff
@@ -26,6 +28,12 @@ import (
+var MaxDuration = (5 * time.Minute)
+
+// Max Segment Size in bytes (cap reading HTTP response body at this size)
+var MaxSegSize = int(MaxDuration.Seconds()) * 1024 * 1024 // 1080p30fps video @ 8000Kb/s or 1MiB/s
```
This might be obvious, but could you explain this calculation for my own understanding?

jailuthra replied:
Assuming the biggest video configuration we expect is 1080p30fps @ 8000kbps (i.e. ~ 1MBytes / s) every second of the video will take up `1024*1024` bytes
Edit: Fixed the comment to make it clearer for future reader

yondonfu replied:
Ah right brain fart on my end heh 👍

jailuthra replied:
Thinking more about this, I'll create a `MaxBitrate` variable here to make it easier on the eyes lol.

yondonfu commented on common/util.go:
```diff
@@ -387,3 +395,15 @@ func JoinURL(url, path string) string {
+// Read at most n bytes from an io.Reader
+// error out if the io.Reader had bytes remaining
+func ReadAtMost(r io.Reader, n int) ([]byte, error) {
+	limitedReader := io.LimitReader(r, int64(n)+1)
```
Should the limit be n as opposed to n + 1?

jailuthra replied:
If we keep it as `n`, `len(b)` in the subsequent check will always be `<= n`, and we won't be able to find out if the underlying `io.Reader` had more bytes than we wanted to read (i.e. `n`)

I'll add a comment to make this clearer.

yondonfu reviewed (changes requested):
Let's add some small tests for the new limiting behavior?

yondonfu commented:
Latest changes look good to go for me pending the tests for the byte reading limiting behavior.

jailuthra committed:
server: tests for MaxSegSize

jailuthra committed:
CHANGELOG_PENDING: Add entry for #1827

jailuthra commented:
> Latest changes look good to go for me pending the tests for the byte reading limiting behavior.

Added a couple of tests in https://github.com/livepeer/go-livepeer/pull/1827/commits/9d23156dc5aa076b4e8ec0becb5d777eede42eba, rebased the PR and added a changelog entry.

Also created a fork debt issue for making the max segment size configurable.

yondonfu reviewed (approved):
LGTM
